_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,1079 @@
+
+```shell
+$ git clone git@github.com:bnahill/PyCortexMDebug.git
+$ cd PyCortexMDebug
+$ python setup.py install
+```
+
+When you next start gdb, you can source the `svd_gdb.py` script and use it to start inspecting registers. Here's some output for the svd plugin we will use in the examples [below](#svd-usage-example):
+
+```
+(gdb) source cmdebug/svd_gdb.py
+(gdb) svd_load cortex-m4-scb.svd
+(gdb) svd
+Available Peripherals:
+    ...
+	SCB:        System control block
+    ...
+(gdb) svd SCB
+Registers in SCB:
+    ...
+	CFSR_UFSR_BFSR_MMFSR:      524288  Configurable fault status register
+    ...
+(gdb) svd SCB CFSR_UFSR_BFSR_MMFSR
+Fields in SCB CFSR_UFSR_BFSR_MMFSR:
+	IACCVIOL:     0  Instruction access violation flag
+	DACCVIOL:     0  Data access violation flag
+	MUNSTKERR:    0  Memory manager fault on unstacking for a return from exception
+	MSTKERR:      0  Memory manager fault on stacking for exception entry.
+	MLSPERR:      0
+	MMARVALID:    0  Memory Management Fault Address Register (MMAR) valid flag
+	IBUSERR:      1  Instruction bus error
+	PRECISERR:    0  Precise data bus error
+	IMPRECISERR:  0  Imprecise data bus error
+	UNSTKERR:     0  Bus fault on unstacking for a return from exception
+	STKERR:       0  Bus fault on stacking for exception entry
+	LSPERR:       0  Bus fault on floating-point lazy state preservation
+	BFARVALID:    0  Bus Fault Address Register (BFAR) valid flag
+	UNDEFINSTR:   0  Undefined instruction usage fault
+	INVSTATE:     1  Invalid state usage fault
+	INVPC:        0  Invalid PC load usage fault
+	NOCP:         0  No coprocessor usage fault.
+	UNALIGNED:    0  Unaligned access usage fault
+	DIVBYZERO:    0  Divide by zero usage fault
+```
+
+### Postmortem Analysis
+
+The previous two approaches are only helpful if we have a debug or physical connection to the device. Once the product has shipped and is out in the field these strategies will not help to triage what went wrong on devices.
+
+One approach is to simply try and reproduce the issue on site. This is a guessing game (are you actually reproducing the customer hit?), can be a huge time sync and in some cases is not even particularly feasible[^14].
+
+Another strategy is to log the fault register and stack values to persistent storage and periocially collect or push the error logs. On the server side, the register values can be decoded and addresses can be symbolicated to try to root cause the crash.
+
+Alternatively an end-to-end firmware error analysis system, such as [Memfault](https://memfault.com/features/error-analysis.html?utm_source=interrupt&utm_medium=link&utm_campaign=cortex-m-faults), can be used to automatically collect, transport, deduplicate and surface the biggest issues in the field. Here is some example output from Memfault for the bad memory read example we will walk through [below](#bad-address-read-example):
```

Comment:
"the biggest issues" sounds hand-wavy. It's a technical post, so I'd going ahead and say we surface faults and exceptions, or something more technical.